rcore-fs-ramfs/src/lib.rs

```diff
@@ -361,6 +361,40 @@ impl INode for LockedINode {
         }
     }
 
+    fn iterate_dir(&self, ctx: &mut DirentWriterContext) -> Result<usize> {
+        let file = self.0.read();
+        if file.extra.type_ != FileType::Dir {
+            return Err(FsError::NotDir);
+        }
+        let idx = ctx.pos();
+        let mut total_written_len = 0;
+        let self_and_parent = {
+            let mut self_and_parent = BTreeMap::new();
+            self_and_parent.insert(String::from("."), file.this.upgrade().unwrap());
+            self_and_parent.insert(String::from(".."), file.parent.upgrade().unwrap());
+            self_and_parent
+        };
```

Comment:
While I like the simplicity of the implementation, I was hoping that the overhead of constructing the `BTreeMap` (at least three heap allocations) can be avoided. So I make the attempt to write an equally-readable, yet more efficient version. The code is meant to show an alternative. It is your call to decide which approach to use :)

```rust
    fn iterate_dir(&self, ctx: &mut DirentWriterContext) -> Result<usize> {
        let file = self.0.read();
        if file.extra.type_ != FileType::Dir {
            return Err(FsError::NotDir);
        }

        macro_rules! write_entry {
            ($ctx:expr, $name:expr, $inode:expr, $total_written:expr) => {
                let ctx = $ctx;
                let name = $name;
                let inode = $inode;
                let total_written = $total_written;

                let entry_file = inode.0.read();
                match ctx.write_entry(
                    name,
                    entry_file.extra.inode as u64,
                    entry_file.extra.type_,
                ) {
                    Ok(written_len) => {
                        *total_written += written_len;
                    }
                    Err(e) => {
                        if *total_written_len > 0 {
                            return Ok(*total_written);
                        } else {
                            Err(e)
                        }
                    }
                }
            }
        }

        let mut total_written = 0;
        let idx = ctx.pos();
        // Write the two special entries
        if idx == 0 {
            let this_inode = file.this.upgrade().unwrap();
            write_entry!(ctx, ".", &this_inode, &mut total_written);
        }
        if idx <= 1 {
            let parent_inode = file.this.upgrade().unwrap();
            write_entry!(ctx, ".", &parent_inode, &mut total_written);
        }
        // Write the normal entries
        let skipped_children = if idx < 2 {
            0
        } else {
            idx - 2
        };
        for (name, inode) in file.children.iter().skip(skipped_children) {
            write_entry!(ctx, name, inode, &mut total_written)?;
        }

        Ok(total_written)
    }
```
(Warning: I haven't compiled or tested the code snippet.)